fop/fom.c

```diff
@@ -813,25 +819,31 @@ static void fom_exec(struct m0_fom *fom)
 	} else {
 		struct m0_fom_callback *cb;
 
-		fom_wait(fom);
-		/*
-		 * If there are pending call-backs, execute them, until one of
-		 * them wakes the fom up. Don't bother to optimize moving
-		 * between queues: this is a rare case.
-		 *
-		 * Note: call-backs are executed in LIFO order.
-		 */
-		M0_ADDB2_PUSH(M0_AVI_FOM_CB);
-		while ((cb = fom->fo_pending) != NULL) {
-			fom->fo_pending = cb_next(cb);
-			cb_run(cb);
+		if (fom->fo_hung_cnt < HUNG_FOP_HIT_MAX_CNT) {
+			fom_wait(fom);
 			/*
-			 * call-back is not allowed to destroy a fom.
+		 	 * If there are pending call-backs, execute them, until one of
+			 * them wakes the fom up. Don't bother to optimize moving
+			 * between queues: this is a rare case.
+			 *
+			 * Note: call-backs are executed in LIFO order.
 			 */
-			M0_ASSERT(m0_fom_phase(fom) != M0_FOM_PHASE_FINISH);
-			if (fom_state(fom) != M0_FOS_WAITING)
-				break;
+			M0_ADDB2_PUSH(M0_AVI_FOM_CB);
+			while ((cb = fom->fo_pending) != NULL) {
+				fom->fo_pending = cb_next(cb);
+				cb_run(cb);
+				/*
+				 * call-back is not allowed to destroy a fom.
+				 */
+				M0_ASSERT(m0_fom_phase(fom) != M0_FOM_PHASE_FINISH);
+				if (fom_state(fom) != M0_FOS_WAITING)
+					break;
+			}
 		}
+
+		if (fom->fo_hung_cnt >= HUNG_FOP_HIT_MAX_CNT)
+			m0_fom_phase_move(fom, M0_ERR(-EAGAIN), M0_FOPH_FAILURE);
```

Comment:
But I have a concern here:
Some of FOM state are not allowed to move to M0_FOPH_FAILURE.
This is configured in fom sm transition table.

For example, the following sate:
```
        [M0_FOPH_TXN_COMMIT_WAIT] = {
                .sd_name      = "tx_commit_wait",
                .sd_allowed   = M0_BITS(M0_FOPH_TXN_INIT, M0_FOPH_FINISH)
        },
```

and this one:
```
        [FOPH_RM_REQ_WAIT] = {
                .sd_name    = "RM_Request_Wait",
                .sd_allowed = M0_BITS(FOPH_RM_REQ_FINISH)
        },
```

And these two fom states were reported to be HUNG in the following ticket:
```
https://jts.seagate.com/browse/COSTOR-898?focusedCommentId=1455506&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-1455506
```

So we need to make sure this state transition is allowed.

Reply:
Agree with comment, this was incorrect approach.